### ISSUE TITLE
Open new window/tab with JavaScript fails

### ISSUE THREAD

**Issue by uhensler** [SafeExamBrowser/seb-mac#230]
When opening a new window/tab with JavaScript an exception occurs because the JavaScript variable `targetWindow` is always `null`.

```
var targetWindow = window.open("","_blank");
...
targetWindow.location.href = '<the real url>';
```

In the GUI this leads to a popup with the message "Quit Safe Exam Browser: Are you sure you want to quit Safe Exam Browser?".

**Comment by uhensler** [comment]
Test page will be sent by private email.

**Comment by danschlet** [comment]
window.open() in SEB for macOS/iOS currently doesn't return the window object when the modern WebView is used. It might work with the classic webView on Mac, at least in all the years no one complained about it. Or maybe people just refrained from using JS to open new windows.

If you need the window object, your website code needs to get it when the new page is loaded in the new window. SEB also doesn't support changing the browser window size or properties, browser windows are always created according to settings in Preferences/Browser ("New browser window size and position").

SEB is a kiosk browser, so it has some restrictions compared to ordinary web browsers. Web applications wanting to run in the kiosk environment might need some adjustments.

**Comment by danschlet** [comment]
In SEB 3.1 there is actually a crash happening, which is already fixed in latest development builds. This doesn't change the window object issue though.

**Comment by uhensler** [comment]
> If you need the window object, your website code needs to get it when the new page is loaded in the new window.

This does not work for us. We need the window to set the URL and load the page.

**Comment by danschlet** [comment]
Can't you set the URL in the window.open command?

Without knowing your use case, there are usually several ways how to reach a goal or workarounds. I will try to make this work, but I can't guess how complicated it will be (didn't find anything in a quick search). I checked and there's also no window object returned when using window.open() in the classic WebView. That means that 12 years since SEB for macOS exists no one was needing the window object (just saying ;-) ).

**Comment by danschlet** [comment]
Can't you set the URL in the window.open command?

Without knowing your use case, there are usually several ways how to reach a goal or workarounds. I will try to make this work, but I can't guess how complicated it will be (didn't find anything in a quick search). I checked and there's also no window object returned when using window.open() in the classic WebView since SEB 3.0. It worked in SEB <= 2.3.2, so the implementation to be able to use both WebView types broke this.

That means that it probably should be fairly easily possible to make it work also in the modern WebView and fix it in the old one, I'll try to add the fix in SEB 3.1.1 (if too time consuming in 3.2).

**Comment by danschlet** [comment]
This is now fixed in [SEB 3.1.1rc (release candidate, please test)](https://github.com/SafeExamBrowser/seb-mac/releases/tag/3.1.1rc). 

I also investigated the same fix for the iOS/iPadOS version of SEB, but the according WKWebView API seems to be broken on iOS. I will file a bug with Apple, without them fixing it I don't think I can do anything to make it work on iOS/iPadOS.
